keps/sig-instrumentation/20190605-metrics-validation-and-verification.md

```diff
@@ -0,0 +1,180 @@
+
+ ### Goals
+
+* Given a stable metric, validate that we cannot remove or modify it (other than adding deprecation information).
+* Given a deprecated but stable metric, validate that we cannot remove or modify it until the deprecation period has elapsed.
+* Given an alpha metric which is promoted to be 'stable', automatically include proper instrumentation reviewers (for schema validation and conformance to metrics guidelines).
+
+ ### Non-Goals
+
+ * Conformance testing will not apply to alpha metrics, since alpha metrics do not have stability guarantees.
+
+ ## Proposal
+
+ We will provide validation for metrics under the [new framework](https://github.com/kubernetes/enhancements/blob/master/keps/sig-instrumentation/20190404-kubernetes-control-plane-metrics-stability.md) with static analysis.
+
+ ## Design Details
+
+ Metrics conformance testing will work in a similar (but not identical) fashion to the generic Kubernetes conformance tests.  Sig-instrumentation will own a directory under `test/instrumentation`. There will be a subdirectory `testdata` in which a file `stable-metrics-list.txt` will live.  This file will be owned by sig-instrumentation. Metrics conformance tests will involve a static analysis script which will traverse the entire codebase and look for metrics which are annotated as 'stable'. For each stable metrics, this script will generate a stringified version of metric metadata (i.e. name, type, labels) which will then be appended together in lexographic order. This will be the output of this script.
+
+ We will add a pre-submit check, which will run in our CI pipeline, which will run our script with the current changes and compare that to existing, committed file. If there is a difference, the pre-submit check will fail. In order to pass the pre-submit check, the original submitter of the PR will have to run a script `test/instrumentation/update-stable-metrics.sh` which will run our static analysis code and overwrite `stable-metrics-list.txt`. This will cause `sig-instrumentation` to be tagged for approvals on the PR (since they own that file).
+ 
+ 
+ ### Static Analysis
+ 
+Similarly to conformance test proposed analysis will be performed on golang source code using default abstract syntaxt tree parser `go/ast`. Handling all possible cases for how metrics can be instantiated would require executing the code itself and is not practical. To reduce number of cases needed handled we will be making assumption of non malicious intent of contributors. There are just to many ways on golang struct can be instantiated that would hide potential stable metrics. To ensure correctness and code simplicity we will be restricting how stable metrics can be declared to one format described below. Alpha metrics will not restricted in the same way to allow flexibility and their dynamic generation.
```

Comment:
We should also note that alpha metrics don't need to be analyzed at all since there are no stability guarantees. That's the driving reason why one could do whatever they wanted with alpha metrics.

Reply:
Changed last sentence to "Alpha metrics will not be analyzed as they don't have any stability guarantees. Their declaration will not have any restrictions, thus allowing dynamic generation."